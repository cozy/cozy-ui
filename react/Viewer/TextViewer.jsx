import React from 'react'
import ReactMarkdown from 'react-markdown'
import PropTypes from 'prop-types'
import cx from 'classnames'
import { withClient, models } from 'cozy-client'
import Spinner from '../Spinner'
import withFileUrl from './withFileUrl'
import styles from './styles.styl'
import NoViewer from './NoViewer'
import { FileDoctype } from '../proptypes'

const MarkdownRenderer = ({ text }) => (
  <ReactMarkdown
    className={cx(styles['viewer-textviewer-content'], 'u-p-1')}
    source={text}
  />
)

const PlainTextRenderer = ({ text }) => (
  <pre
    className={cx(styles['viewer-textviewer-content'], 'u-mh-auto', 'u-mv-2')}
  >
    {text}
  </pre>
)

const Loader = () => (
  <div className={styles['viewer-textviewer']}>
    <Spinner size="xxlarge" middle noMargin color="white" />
  </div>
)

export const isMarkdown = file =>
  file.mime === 'text/markdown' ||
  /.md$/.test(file.name) ||
  models.file.isNote(file)
export class TextViewer extends React.Component {
  state = {
    text: '',
    isMarkdown: false,
    loading: true,
    error: null
  }
  _mounted = false

  componentDidMount() {
    this._mounted = true
    this.loadFile()
  }

  componentWillUnmount() {
    this._mounted = false
  }

  async loadFile() {
    const { url, file } = this.props
    try {
      const parsedURL = new URL(url)
      const client = this.props.client.getStackClient()
      const response = await client.fetch('GET', parsedURL.pathname)
      const text = await response.text()

      if (this._mounted) {
        this.setState({
          text,
          isMarkdown: isMarkdown(file),
          loading: false
        })
      }
    } catch (error) {
      // eslint-disable-next-line no-console
      console.warn(error)
      if (this._mounted) {
        this.setState({
          loading: false,
          error
        })
      }
    }
  }

  render() {
    const { loading, error, text, isMarkdown } = this.state
    const { file, renderFallbackExtraContent } = this.props
    if (loading) return <Loader />
    else if (error)
      return (
        <NoViewer
          file={file}
          renderFallbackExtraContent={renderFallbackExtraContent}
        />
      )
    else
      return (
        <div className={styles['viewer-textviewer']}>
          <h2 className={cx(styles['viewer-filename'], 'u-mt-3', 'u-mb-1')}>
            {file.name}
          </h2>
          {isMarkdown ? (
            <MarkdownRenderer text={text} />
          ) : (
            <PlainTextRenderer text={text} />
          )}
        </div>
      )
  }
}

TextViewer.propTypes = {
  client: PropTypes.object.isRequired,
  url: PropTypes.string.isRequired,
  file: FileDoctype.isRequired,
  renderFallbackExtraContent: PropTypes.func
}

export default withFileUrl(withClient(TextViewer))
