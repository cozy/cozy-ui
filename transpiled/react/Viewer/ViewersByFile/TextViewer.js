import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _regeneratorRuntime from "@babel/runtime/regenerator";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import cx from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import ReactMarkdown from 'react-markdown';
import { withClient, models } from 'cozy-client';
var styles = {
  "viewer-imageviewer": "styles__viewer-imageviewer___26k0p",
  "viewer-noviewer": "styles__viewer-noviewer___auG-6",
  "viewer-audioviewer": "styles__viewer-audioviewer___1OQPB",
  "viewer-videoviewer": "styles__viewer-videoviewer___NhFoe",
  "viewer-pdfviewer": "styles__viewer-pdfviewer___1gTP9",
  "viewer-textviewer": "styles__viewer-textviewer___3u5Zw",
  "viewer-canceled": "styles__viewer-canceled___pOA_O",
  "viewer-textviewer-content": "styles__viewer-textviewer-content___PB-c3",
  "viewer-filename": "styles__viewer-filename___3jZCt",
  "viewer-pdfviewer-pdf": "styles__viewer-pdfviewer-pdf___16ID9",
  "viewer-pdfviewer-page": "styles__viewer-pdfviewer-page___2RPuw",
  "viewer-pdfviewer-toolbar": "styles__viewer-pdfviewer-toolbar___3NXOk",
  "viewer-pdfMobile": "styles__viewer-pdfMobile___25FPg",
  "viewer-pdfMobile--image": "styles__viewer-pdfMobile--image___3gpFL"
};
import { FileDoctype } from "cozy-ui/transpiled/react/proptypes";
import NoViewer from "cozy-ui/transpiled/react/Viewer/NoViewer";
import ViewerSpinner from "cozy-ui/transpiled/react/Viewer/components/ViewerSpinner";
import { isFileEncrypted } from "cozy-ui/transpiled/react/Viewer/helpers";
import withFileUrl from "cozy-ui/transpiled/react/Viewer/hoc/withFileUrl";

var MarkdownRenderer = function MarkdownRenderer(_ref) {
  var text = _ref.text;
  return /*#__PURE__*/React.createElement(ReactMarkdown, {
    className: cx(styles['viewer-textviewer-content'], 'u-p-1'),
    source: text
  });
};

var PlainTextRenderer = function PlainTextRenderer(_ref2) {
  var text = _ref2.text;
  return /*#__PURE__*/React.createElement("pre", {
    className: cx(styles['viewer-textviewer-content'], 'u-mh-auto', 'u-mv-2')
  }, text);
};

var Loader = function Loader() {
  return /*#__PURE__*/React.createElement("div", {
    className: styles['viewer-textviewer']
  }, /*#__PURE__*/React.createElement(ViewerSpinner, null));
};

export var isMarkdown = function isMarkdown(file) {
  return file.mime === 'text/markdown' || /.md$/.test(file.name) || models.file.isNote(file);
};
export var TextViewer = /*#__PURE__*/function (_React$Component) {
  _inherits(TextViewer, _React$Component);

  var _super = _createSuper(TextViewer);

  function TextViewer() {
    var _this;

    _classCallCheck(this, TextViewer);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "state", {
      text: '',
      isMarkdown: false,
      loading: true,
      error: null
    });

    _defineProperty(_assertThisInitialized(_this), "_mounted", false);

    return _this;
  }

  _createClass(TextViewer, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this._mounted = true;
      this.loadFile();
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this._mounted = false;
    }
  }, {
    key: "loadFile",
    value: function () {
      var _loadFile = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
        var _this$props, url, file, response, _URL, pathname, client, text;

        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _this$props = this.props, url = _this$props.url, file = _this$props.file;
                _context.prev = 1;

                if (!isFileEncrypted(file)) {
                  _context.next = 8;
                  break;
                }

                _context.next = 5;
                return fetch(url);

              case 5:
                response = _context.sent;
                _context.next = 13;
                break;

              case 8:
                _URL = new URL(url), pathname = _URL.pathname;
                client = this.props.client.getStackClient();
                _context.next = 12;
                return client.fetch('GET', pathname);

              case 12:
                response = _context.sent;

              case 13:
                _context.next = 15;
                return response.text();

              case 15:
                text = _context.sent;

                if (this._mounted) {
                  this.setState({
                    text: text,
                    isMarkdown: isMarkdown(file),
                    loading: false
                  });
                }

                _context.next = 23;
                break;

              case 19:
                _context.prev = 19;
                _context.t0 = _context["catch"](1);
                // eslint-disable-next-line no-console
                console.warn(_context.t0);

                if (this._mounted) {
                  this.setState({
                    loading: false,
                    error: _context.t0
                  });
                }

              case 23:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[1, 19]]);
      }));

      function loadFile() {
        return _loadFile.apply(this, arguments);
      }

      return loadFile;
    }()
  }, {
    key: "render",
    value: function render() {
      var _this$state = this.state,
          loading = _this$state.loading,
          error = _this$state.error,
          text = _this$state.text,
          isMarkdown = _this$state.isMarkdown;
      var _this$props2 = this.props,
          file = _this$props2.file,
          renderFallbackExtraContent = _this$props2.renderFallbackExtraContent;
      if (loading) return /*#__PURE__*/React.createElement(Loader, null);else if (error) return /*#__PURE__*/React.createElement(NoViewer, {
        file: file,
        renderFallbackExtraContent: renderFallbackExtraContent
      });else return /*#__PURE__*/React.createElement("div", {
        className: styles['viewer-textviewer']
      }, /*#__PURE__*/React.createElement("h2", {
        className: cx(styles['viewer-filename'], 'u-mt-3', 'u-mb-1')
      }, file.name), isMarkdown ? /*#__PURE__*/React.createElement(MarkdownRenderer, {
        text: text
      }) : /*#__PURE__*/React.createElement(PlainTextRenderer, {
        text: text
      }));
    }
  }]);

  return TextViewer;
}(React.Component);
TextViewer.propTypes = {
  client: PropTypes.object.isRequired,
  url: PropTypes.string.isRequired,
  file: FileDoctype.isRequired,
  renderFallbackExtraContent: PropTypes.func
};
export default withFileUrl(withClient(TextViewer));