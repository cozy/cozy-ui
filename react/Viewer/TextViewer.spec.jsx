import React from 'react'
import { shallow } from 'enzyme'
import { TextViewer, isMarkdown } from './TextViewer'
import { createMockClient } from 'cozy-client'
import renderer from 'react-test-renderer'
const props = {
  client: createMockClient({}),
  url: 'https://foo.mycozy.cloud',
  file: {
    _id: '1',
    _type: 'io.cozy.files',
    name: 'My File'
  }
}

describe('isMarkdown function', () => {
  it('test markdown function', () => {
    const note = {
      name: 'My Note.cozy-note',
      type: 'file',
      metadata: {
        content: 'my prosemirror content',
        schema: '1',
        title: 'prosemirror title',
        version: '3'
      }
    }
    expect(isMarkdown({ mime: 'text/markdown' })).toBe(true)
    expect(isMarkdown({ name: 'text.md' })).toBe(true)
    expect(isMarkdown(note)).toBe(true)
    expect(isMarkdown({ name: 'text.txt' })).toBe(false)
  })
})
describe('TextViewer Component', () => {
  it('should display the loader ', () => {
    const comp = shallow(<TextViewer {...props} />)
    expect(comp).toMatchSnapshot()
  })
  it('should display the error component and render with renderFallback', () => {
    const comp = renderer.create(
      <TextViewer
        {...props}
        renderFallbackExtraContent={file => <span>{file.name}</span>}
      />
    )
    comp.getInstance().setState({ error: true, loading: false })
    expect(comp.toJSON()).toMatchSnapshot()
  })
  it('should display the text viewer', () => {
    const comp = renderer.create(<TextViewer {...props} />)
    comp.getInstance().setState({
      loading: false,
      isMarkdown: false,
      text: 'The content of my file'
    })
    expect(comp.toJSON()).toMatchSnapshot()
  })

  it('should display the markdown viewer', () => {
    const comp = renderer.create(<TextViewer {...props} />)
    comp.getInstance().setState({
      loading: false,
      isMarkdown: true,
      text:
        "It's very easy to make some words **bold** and other words *italic* with Markdown"
    })
    expect(comp.toJSON()).toMatchSnapshot()
  })
})
