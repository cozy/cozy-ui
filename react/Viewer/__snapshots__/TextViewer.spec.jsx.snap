// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextViewer Component should display the error component and render with renderFallback 1`] = `
<div
  className="styles__viewer-noviewer___3heA6"
>
  <svg
    className="styles__icon___23x3R"
    height={140}
    style={Object {}}
    width={160}
  >
    <use
      xlinkHref="#file-type-files"
    />
  </svg>
  <p
    className="styles__viewer-filename___36bWW"
  >
    My File
  </p>
  <span>
    My File
  </span>
</div>
`;

exports[`TextViewer Component should display the loader  1`] = `<Loader />`;

exports[`TextViewer Component should display the markdown viewer 1`] = `
<div
  className="styles__viewer-textviewer___cISlz"
>
  <h2
    className="styles__viewer-filename___36bWW u-mt-3 u-mb-1"
  >
    My File
  </h2>
  <div
    className="styles__viewer-textviewer-content___q2RzR u-p-1"
  >
    <p>
      It's very easy to make some words 
      <strong>
        bold
      </strong>
       and other words 
      <em>
        italic
      </em>
       with Markdown
    </p>
  </div>
</div>
`;

exports[`TextViewer Component should display the text viewer 1`] = `
<div
  className="styles__viewer-textviewer___cISlz"
>
  <h2
    className="styles__viewer-filename___36bWW u-mt-3 u-mb-1"
  >
    My File
  </h2>
  <pre
    className="styles__viewer-textviewer-content___q2RzR u-mh-auto u-mv-2"
  >
    The content of my file
  </pre>
</div>
`;
